builtin/v9/verifreg/verified_registry_state.go

```diff
@@ -63,3 +75,151 @@ func ConstructState(store adt.Store, rootKeyAddress addr.Address) (*State, error
+	if clientIdAddr.Protocol() != address.ID {
+		return nil, false, xerrors.Errorf("can only look up ID addresses")
+	}
+
+	innerHamtCid, err := getInnerHamtCid(store, abi.IdAddrKey(clientIdAddr), st.Allocations)
+	if err != nil {
+		return nil, false, err
+	}
+
+	idToAllocationMap, err := adt.AsMap(store, innerHamtCid, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return nil, false, xerrors.Errorf("couldn't get inner map: %x", err)
+	}
+
+	var allocation Allocation
+	if found, err := idToAllocationMap.Get(allocationId, &allocation); err != nil {
+		return nil, false, xerrors.Errorf("looking up allocation ID: %d: %w", allocationId, err)
+	} else if !found {
+		return nil, false, nil
+	}
+
+	clientId, err := address.IDFromAddress(clientIdAddr)
+	if err != nil {
+		return nil, false, xerrors.Errorf("couldn't get ID from clientIdAddr: %s", clientIdAddr)
+	}
+
+	if uint64(allocation.Client) != clientId {
+		return nil, false, xerrors.Errorf("clientId: %d did not match client in allocation: %d", clientId, allocation.Client)
+	}
+
+	return &allocation, true, nil
+}
+
+func (st *State) FindClaim(store adt.Store, providerIdAddr address.Address, claimId ClaimId) (*Claim, bool, error) {
+	if providerIdAddr.Protocol() != address.ID {
+		return nil, false, xerrors.Errorf("can only look up ID addresses")
+	}
+
+	innerHamtCid, err := getInnerHamtCid(store, abi.IdAddrKey(providerIdAddr), st.Claims)
+	if err != nil {
+		return nil, false, err
+	}
+
+	idToClaimsMap, err := adt.AsMap(store, innerHamtCid, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return nil, false, xerrors.Errorf("couldn't get inner map: %x", err)
+	}
+
+	var claim Claim
+	if found, err := idToClaimsMap.Get(claimId, &claim); err != nil {
+		return nil, false, xerrors.Errorf("looking up allocation ID: %d: %w", claimId, err)
+	} else if !found {
+		return nil, false, nil
+	}
+
+	providerId, err := address.IDFromAddress(providerIdAddr)
+	if err != nil {
+		return nil, false, xerrors.Errorf("couldn't get ID from providerIdAddr: %s", providerIdAddr)
+	}
+
+	if uint64(claim.Provider) != providerId {
+		return nil, false, xerrors.Errorf("providerId: %d did not match provider in claim: %d", providerId, claim.Provider)
+	}
+
+	return &claim, true, nil
+}
+
+func getInnerHamtCid(store adt.Store, key abi.Keyer, mapCid cid.Cid) (cid.Cid, error) {
+	actorToHamtMap, err := adt.AsMap(store, mapCid, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("couldn't get outer map: %x", err)
+	}
+
+	var innerHamtCid cbg.CborCid
+	if found, err := actorToHamtMap.Get(key, &innerHamtCid); err != nil {
+		return cid.Undef, xerrors.Errorf("looking up key: %s: %w", key, err)
+	} else if !found {
+		return cid.Undef, xerrors.Errorf("did not find key: %s", key)
+	}
+
+	return cid.Cid(innerHamtCid), nil
+}
+
+func (st *State) AllocationsMap(store adt.Store, clientIdAddr address.Address) (map[AllocationId]Allocation, error) {
```

Comment:
Nit: suggest AllocationsAsMap, or LoadAllocationsToMap or something else that suggests the amount of work this will do